### ISSUE TITLE
Hybrid: connection lost after editing appsody project

### ISSUE THREAD

**Issue by jopit** [eclipse/codewind#1472]
I created a remote appsody node project.  I then changed the 'app.js' file and saved it.  While it was building, the connection to the IDE was lost.  I saw this with both VSCode and Eclipse, and can reproduce it fairly reliably.  Refreshing the connection will often re-establish the connection.

Here is the pfe log
[pfe.txt](https://github.com/eclipse/codewind/files/3947472/pfe.txt)

Here's the relevant part of the vscode log
```
[ms-vscode.vscode-typescript-tslint-plugin] Accessing a window scoped configuration for a resource is not expected. To associate 'tslint.jsEnable' to a resource, define its scope to 'resource' in configuration contributions in 'package.json'.
extensionHostProcess.js:671
[ms-vscode.vscode-typescript-tslint-plugin] Accessing a window scoped configuration for a resource is not expected. To associate 'tslint.jsEnable' to a resource, define its scope to 'resource' in configuration contributions in 'package.json'.
extensionHostProcess.js:671
[DBUG: 17:04:14.335 Connection.js:81]: Codewind on Fyre onDisconnect
Logger.js:140
[DBUG: 17:04:14.336 Connection.js.setState():103]: Codewind on Fyre is now Network Error
Logger.js:140
[DBUG: 17:04:14.337 Connection.js:89]: Codewind on Fyre is now disconnected
Logger.js:140
[DBUG: 17:04:56.584 Connection.js:81]: Codewind on Fyre onDisconnect
Logger.js:140
[DBUG: 17:05:05.771 Connection.js:81]: Codewind on Fyre onDisconnect
Logger.js:140
[DBUG: 17:05:10.229 Connection.js:63]: Codewind on Fyre onConnect
Logger.js:140
[DBUG: 17:05:10.230 Connection.js:73]: Codewind on Fyre is now connected
Logger.js:140
[DBUG: 17:05:10.231 Connection.js:75]: Codewind on Fyre is now ready - initialize finished before connect
Logger.js:140
[DBUG: 17:05:10.232 Connection.js.setState():103]: Codewind on Fyre is now Ready
Logger.js:140
[DBUG: 17:05:11.458 Connection.js:81]: Codewind on Fyre onDisconnect
Logger.js:140
[DBUG: 17:05:11.459 Connection.js.setState():103]: Codewind on Fyre is now Network Error
Logger.js:140
[DBUG: 17:05:11.460 Connection.js:89]: Codewind on Fyre is now disconnected
```

**Comment by jopit** [comment]
I don't see the same problem with a codewind-style node project.

**Comment by tetchel** [comment]
I wonder if it may be because the code is blocked on some IO which is causing the repeated error in the logs;

`[33m[10/12/19 22:04:26 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN`

**Comment by jopit** [comment]
Making this stopship because it looks really bad

**Comment by jopit** [comment]
I've reproduced this on two different OpenShift clusters, one of them freshly created, so it's not a system-specific problem.

**Comment by malincoln** [comment]
@sghung @makandre pls review.  Thanks

**Comment by makandre** [comment]
I tried a few times and reproduced it once... doesn't always happen it would seem.

It may have something to do with the app log being open; I somehow got into a state where the log view says the log is disconnected and I can't close the log.  Once it got to that state the disconnect happens on the edits.  Otherwise I was able to do multiple edits and the connection didn't get cut

**Comment by makandre** [comment]
... happened again on a new project, this time didn't have the log open :(

**Comment by makandre** [comment]
I also see a lot of defunct processes in PFE..

```
root      58522    154  0 14:56 ?        00:00:00 [rm] <defunct>
```

Taken together with the `EAGAIN` error, it seems to suggest something is starting up a lot of (`rm`?) processes and maybe reaching the max limit of process you could have (googling `EAGAIN` suggest this may indicate too many processes)

There's a `rm` call in https://github.com/eclipse/codewind/blob/master/src/pfe/portal/modules/utils/sharedFunctions.js#L142

**Comment by makandre** [comment]
[here](https://github.com/eclipse/codewind/blob/master/src/pfe/portal/routes/projects/remoteBind.route.js#L229) and [here](https://github.com/eclipse/codewind/blob/master/src/pfe/portal/routes/projects/remoteBind.route.js#L242) are places where we are potentially calling `rm` on a lot of files

**Comment by makandre** [comment]
Just prior to the `EAGAIN`, it is trying to remove `node_modules` file by file, hence all the `rm` processes are overwhelming the process limit.

```
[32m[12/12/19 15:39:07 /portal/routes/projects/remoteBind.route.js] [INFO][39m Removing locally deleted files from project: rm-apnode04, ID: 2265e5d0-1cf5-11ea-a882-1da54a0945e7 - node_modules/lodash/_Promise.js, node_modules/lodash/_Set.js, node_modules/lodash/_SetCache.js, node_modules/lodash/_Stack.js, 
...
```

For appsody projects, the project dir is still mounted to the app container, hence when the app container does `npm install`, the `node_modules` it lays down will appear inside PFE's copy of the project.  But the project on user's local disk does not have `node_modules`, this causes PFE to consider those files as deleted and it will try to delete them.

Now it's a matter of figuring out how to make PFE ignore `node_modules`; not sure if `ignoredPath` is applicable here or if it's something we can do in `cwctl project sync`

@jgwest maybe you have some insight?

**Comment by james-wallis** [comment]
I've had a look through the logs provided:
* agree with @makandre that its the `deleteFilesInArray` file which is causing problems with Appsody (as its going file by file).
* concerned that it appears @jopit has only a single project in PFE.
* I think the below lines show that PFE is being blocked for over 30 seconds while running that function (very bad if true).
```
33m[10/12/19 22:04:26 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN
[33m[10/12/19 22:04:26 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN
[33m[10/12/19 22:04:26 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN
[32m[10/12/19 22:04:26 socketAuth.js] [INFO][39m UISocket : [disconnect] - /default#9eK27KYJvWqOvcBXAAAX disconnected.
[32m[10/12/19 22:04:26 LoadRunner.js] [INFO][39m Loadrunner has disconnected
[33m[10/12/19 22:05:02 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN
[33m[10/12/19 22:05:02 sharedFunctions.js] [WARN][39m spawn /bin/sh EAGAIN
```

IMO the only way to fix that is to add functionality that enables PFE to know what directories have been deleted rather than just files - related to https://github.com/eclipse/codewind/pull/1310/ (which noticed a directory isn't being deleted). Then PFE can delete directories if the all of its contents are removed rather than going file by file as is the current behaviour.

I am going to try to:
- [ ] Recreate this locally (can't see any reason why this would be remote only
- [ ] In remote if I cannot do locally
- [ ] In version 0.6.1 to check for regression

**Comment by makandre** [comment]
But in appsody case, we do not want to delete the node_modules directory. Doing so will probably break the running application; remember that the project folder is mounted to the appsody app container

**Comment by james-wallis** [comment]
Ok, so PFE is deleting the files which we are not told to keep and as locally the node_modules folder will not exist it is also being deleted.
PFE needs to read the projects `ignoredFiles` before deleting things?

**Comment by jopit** [comment]
To confirm, I did only have the one project in PFE.

**Comment by makandre** [comment]
@james-wallis yes perhaps somewhere around https://github.com/eclipse/codewind/blob/master/src/pfe/portal/routes/projects/remoteBind.route.js#L222 when PFE is computing what to delete, it needs to ignore `node_modules` dir.

But I think we need to consider the following:

- the appsody project handler does not define any `defaultIgnoredPath`, since everything* in an appsody project is supposed to be synced over to PFE.  It is then up to appsody CLI to mount that project folder and monitor for file changes (appsody knows that to ignore for a given appsody stack)

- that said, I do see some known file and folders are by default ignored and not synced over to PFE. That's defined in `cwctl` https://github.com/eclipse/codewind-installer/blob/master/pkg/project/sync.go#L249

- I guess the bottomline is, we know certain files and folders (e.g. `node_modules`) won't ever be synced over to PFE, but if such files and folders do appear in PFE's copy of the project, PFE shouldn't go and remove them

**Comment by makandre** [comment]
Aside from not removing `node_modules` for this case, the fix to delete by dir rather than file-by-file is also a fix we need regardless.

**Comment by makandre** [comment]
Now that I think about it, I think this delete behaviour is actually a larger problem for appsody.

Basically, any runtime resources that the appsody app container creates in the project folder, they will be in PFE's copy of the project but not in user's local copy of the project.  If this condition causes PFE to attempt to delete those resources, then it's a real problem.

We need to somehow distinguish between the above conditions vs. a real deletion of files by the user in their local project.

**Comment by james-wallis** [comment]
**End of Friday status**
Had problems all day trying to recreate this issue in a remote scenario. 
I was able to create a project using the master branch of VSCode but it was not reflecting my file changes in a remote deployment only (locally was fine). Tried updating everything to master + re-installing on OpenShift as I wasn't aware anyone else was seeing an issue with file changes, still didn't work. Finally I dropped to just `cwctl` and saw issues with the `sync` command although it started working nearing the end of the day. 
No one else has seen any issues with `sync` so its likely user error or versions not matching up.

Plan now/on Monday is to continue recreating this bug.

**Comment by jopit** [comment]
I can assure you that my versions do match up.

**Comment by james-wallis** [comment]
@jopit I meant the reasons for my Remote Codewind not working was the versions.

**Comment by jopit** [comment]
@james-wallis Oh, ok, my apologies, I misunderstood your comment.

**Comment by james-wallis** [comment]
I've now recreated this issue in OpenShift (using remote) and have found that we have different behaviours for Appsody locally (Docker) and in Kube. Essentially in Docker we've not seen this issue as the PFE container does not see the `node_modules` directory.

**Local Docker**
* Appsody projects go into the `mounted-workspace` not `codewind-workspace`
* In PFE the projects `node_modules` directory is empty.
* The project container itself has a volume mount for its `node_modules` directory so its contents are in a Docker volume (not PFE or anything PFE controls).

This means that when PFE does its file deletion on `sync`, the `node_modules` is unaffected as PFE sees it as empty.

**Remote Kubernetes**
* Appsody projects go into `codewind-workspace`
* `codewind-workspace/{project}` in PFE has a populated `node_modules`
* App container also has `node_modules`
  * App containers `node_modules` is not bound to PFE and so adding a file into here will not show up in PFE.

This means that when PFE does its file deletion on `sync`, the `node_modules` has its contents deleted as PFE can see it in its entirety.

There is no logic in PFE to not delete `node_modules` or `.odo` projects (another issue) as PFE is given a `fileList` with the `sync` API which it uses as a keep file list. In other words, if the projects directory inside PFE contains files that are not present in the keep file list, they will be deleted. So as the users local directory does not contain the `node_modules` folder, it is not in the keep file list and is deleted.

**Comment by malincoln** [comment]
Per discussion with @tobespc on checkpoint call, this is related to 1514, 1520, and 1533.
James should land a fix today for 0.7.0.  @tobespc pls confirm defect list.

**Comment by makandre** [comment]
@james-wallis yes that perfectly explains why we don't see this on local.

Re:

> Although we're currently deleting a projects node_modules directory in Kube, the project does rebuild correctly. It is only PFE which disconnects, the project is rebuilt before I can connect to PFE.

Maybe it rebuilds correctly because the deletion of `node_module` actually fails?  In any case, we need a solution that won't delete the `node_module` folder.

**Comment by james-wallis** [comment]
Discussed with @tobespc and @jcockbain.

**0.7.0**
Add `node_modules` and `.odo` to a specific ignore list when deleting files within `codewind-workspace`. Basically just an `if node_modules don't delete`.

**Master**
 * Add a `directoryList` to the `sync`/`upload/end` API so that we delete by directory where possible not by files (fixes `node_modules` being deleted file by file).
* Add `.odo` to a protected list via the extension see @jcockbain comment  https://github.com/eclipse/codewind/issues/1533#issuecomment-566093070
    * Appsody extension could also add `node_modules` to be protected this way.

**Comment by makandre** [comment]
The only concern I have with the protected list approach is that we don't necessarily know all the appsody stacks that would be used with Codewind, so any protected list provided by the extension would need upkeep and only be current on a "best-effort" basis (I would imagine it would be similar for ODO as well? cc: @GeekArthur )

To avoid that we would need a way to distinguish files that are uploaded by sync vs files generated in PFE during build/run.  I'm wondering if this can be achieved by adding some extended file attributes to files that are synced to PFE?

@james-wallis @jcockbain

**Comment by tobespc** [comment]
same issue as #1533

**Comment by GeekArthur** [comment]
My concern is `.odo` and `node_modules` are two different level directory/file ignore, `.odo` is project type level which applies to all odo projects (shouldn't delete `.odo` for all odo projects, project language doesn't affect), but `node_modules` is project language level which applies to all projects (Codewind, Appsody and odo project types) with nodejs language (I believe Codewind project type nodejs language already added `node_modules` into `.cw-settings` file then it doesn't have this issue?). So if `ignoredPaths` in `.cw-settings` has the capability of ignoring sync, we can detect both project type and project language when we create/import project then add corresponding ignore directory/file into the `.cw-settings`. For example, if we create odo nodejs project, we can add `.odo` (project type level) and `node_modules` (project language level) into its `.cw-settings` to ignore sync.

**Comment by james-wallis** [comment]
Agree that `.odo` and `node_modules` should be handled differently.

I'd still make the case that it would be a good enhancement for this project if an extension could tell PFE a list of paths to ignore (don't delete/modify). The ODO extension can use this to tell PFE to ignore its `.odo` directory as it is very extension specific. This would fit inside the `codewind.yaml` - PFE can read it when an extension is initialised.

Regarding the `node_modules` front, I'm not sure how to handle it as it only happens with an Appsody project. I do understand @makandre's point that Appsody can change and add templates without us knowing making an `ignoredList` difficult to upkeep, however we'll have to keep the ignoredList up to date if we contain it inside PFE. My next piece of work is to delete by directories not just by file so that will "fix" the `node_modules` deletion anyway (it'll still get deleted but Codewind shouldn't crash from overloading). 

I do not think that we should use the `.cw-settings` file for this as I don't think the user should be in control of it, especially in the case of `.odo` where it cannot recover if it is deleted.

**Comment by makandre** [comment]
At present it may only be appsody projects that is exhibiting the problem with `node_modules`, but I think the general issue here, where files generated in PFE's copy of the project during build/run is deleted, can potentially happen for any project type (the problem with `.odo` certainly falls into this category as well)

Totally agree that having to maintain a list of protected/ignored resources, whether by the extension or in PFE, is not that ideal.  That's why I was wondering if we can mark files that are synced to PFE with some kind of extended file attribute (admittedly, my knowledge about extended file attributes only comes from a quick google search which led me to the description on https://en.wikipedia.org/wiki/Extended_file_attributes#Linux; so I have no idea if this would work or not)

But assuming we could somehow mark a file that is synced to PFE, then we encounter a file that exists in PFE but not on the user's local disk, we could do this:

```
if (file is marked meaning it was copied from user disk) {
    delete it
} else {
    leave it alone
}
```

**Comment by GeekArthur** [comment]
@james-wallis 
> I do not think that we should use the .cw-settings file for this as I don't think the user should be in control of it, especially in the case of .odo where it cannot recover if it is deleted.
That makes sense, currently only `.odo` is generated on PFE side and it's the only directory in the protected list for odo projects as odo has its own sync mechanism from PFE to build container, that means all generated build files except `.odo` will be in build container but not in PFE so they don't have this issue. Currently odo `codewind.yaml` file already had `.odo` as detection here: https://github.com/eclipse/codewind-odo-extension/blob/master/codewind.yaml#L6, Portal may directly use that for projected list or we can design a new field for extension's protected list in `codewind.yaml`. Either way, please let me know then I can handle this on odo side for long-term fix.

**Comment by tobespc** [comment]
issue 1533 has a fix that will fix this for 0.7.0

However, there are other improvements to be made in the master branch going forwards.  Targetting this at 0.8.0

**Comment by malincoln** [comment]
@jopit pls verify and close.  Thanks

**Comment by jopit** [comment]
Verified
